ryhope/src/storage/memory.rs

```diff
@@ -286,53 +357,218 @@ where
+            .collect_vec();
+        to_be_erased_epochs.into_iter().for_each(|epoch| {
+            self.0.remove(&epoch);
+        });
+    }
+
+    fn add_epoch(
+        &mut self,
+        user_epoch: UserEpoch,
+        incremental_epoch: IncrementalEpoch,
+    ) -> Result<()> {
+        // double check that we are either replacing an existing `IncrementalEpoch`
+        // in the map or we are adding the next incremental one. This check ensures
+        // that `IncrementalEpoch`s found in the map are always incremental
+        let num_epochs = self.0.len();
         ensure!(
-            epoch <= self.current_epoch(),
-            "unable to rollback to epoch `{}` more recent than current epoch `{}`",
-            epoch,
-            self.current_epoch()
+            incremental_epoch as usize <= num_epochs,
+            "Inserted IncrementalEpoch is too big: found {incremental_epoch}, maximum is {num_epochs}"
         );
+        // check that the `user_epoch` being added is associated to the correct
+        // `incremental_epoch`, according to the ordering in the map
+        if let Some(smaller_epoch) = self.try_to_user_epoch_inner(incremental_epoch - 1) {
+            ensure!(user_epoch > smaller_epoch);
+        }
+        if let Some(bigger_epoch) = self.try_to_user_epoch_inner(incremental_epoch + 1) {
+            ensure!(user_epoch < bigger_epoch)
+        }
+        // if we are replacing an existing `IncrementalEpoch`, ensure that
+        // we remove the old mapping entry
+        if let Some(epoch) = self.try_to_user_epoch_inner(incremental_epoch) {
+            self.0.remove(&epoch);
+        }
 
-        self.mem.truncate((epoch + 1).try_into().unwrap());
-
+        self.0.insert(user_epoch, incremental_epoch);
         Ok(())
     }
 }
 
+impl EpochMapper for InMemoryEpochMapper {
+    async fn try_to_incremental_epoch(&self, epoch: UserEpoch) -> Option<IncrementalEpoch> {
+        self.try_to_incremental_epoch_inner(epoch)
+    }
+
+    async fn try_to_user_epoch(&self, epoch: IncrementalEpoch) -> Option<UserEpoch> {
+        self.try_to_user_epoch_inner(epoch)
+    }
+
+    async fn add_epoch_map(
+        &mut self,
+        user_epoch: UserEpoch,
+        incremental_epoch: IncrementalEpoch,
+    ) -> Result<()> {
+        self.add_epoch(user_epoch, incremental_epoch)
+    }
+}
+
 /// A RAM-backed storage for tree data.
-pub struct InMemory<T: TreeTopology, V: Debug + Sync> {
+pub struct InMemory<T: TreeTopology, V: Clone + Debug + Send + Sync, const READ_ONLY: bool> {
     /// Storage for tree state.
     state: VersionedStorage<<T as TreeTopology>::State>,
     /// Storage for topological data.
     nodes: VersionedKvStorage<<T as TreeTopology>::Key, <T as TreeTopology>::Node>,
     /// Storage for node-associated data.
     data: VersionedKvStorage<<T as TreeTopology>::Key, V>,
+    epoch_mapper: SharedEpochMapper<InMemoryEpochMapper, READ_ONLY>,
```

Comment:
Please comment this field, especially in which direction it maps.

Reply:
Done in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/425/commits/23affb5f3b0c62fc1ab825043196a2f89e20039c